pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_canceler.go

```diff
@@ -86,21 +99,31 @@ func (w *stuckRangeFeedCanceler) ping() {
 		// ping() event arrives at 29.999s, the timer should only fire
 		// at 90s, not 60s.
 		w.t = time.AfterFunc(3*threshold/2, func() {
-			// NB: important to store _stuck before canceling, since we
-			// want the caller to be able to detect stuck() after ctx
-			// cancels.
-			atomic.StoreInt32(&w._stuck, 1)
-			w.cancel()
+			// NB: trigger cancellation only if currently active.
+			if atomic.CompareAndSwapInt32(&w._state, int32(active), int32(stuck)) {
+				w.cancel()
+			}
 		})
 		w.resetTimerAfter = timeutil.Now().Add(threshold / 2)
 	}
 
+	if !atomic.CompareAndSwapInt32(&w._state, int32(inactive), int32(active)) {
```

Comment:
We're adding in three atomic operations in the hot path here. Do we need to benchmark the overhead of these, or is it likely negligible once we factor in IO and such?

Reply:
I think it's negligable.